python/hail/table.py

```diff
@@ -1680,6 +1680,8 @@ def join(self, right, how='inner') -> 'Table':
             Joined table.
 
         """
+        if not self.key.dtype == right.key.dtype:
+            raise ValueError(f"'join': key mismatch:\n  Left:  '{self.key.dtype}'\n  Right: '{right.key.dtype}'")
```

Comment:
I think you only want to check the field types are the same. right now it's testing struct equality including field names.